traits/trait_set_object.py

```diff
@@ -51,213 +50,518 @@ def __repr__(self):
+
+            notifier(removed, added)
+
+        Where 'added' is a set containing new values that have been added.
+        And 'removed' is a set containing old values that have been removed.
+
+        If this argument is not given, the list of notifiers is initially
+        empty.
+
+    Attributes
+    ----------
+    item_validator : callable
+        Called to validate and/or transform items added to the set. The
+        callable should accept a single item and return the transformed
+        item, raising TraitError for invalid items.
+    notifiers : list of callable
+        A list of callables with the signature::
+
+            notifier(removed, added)
+
+        Where 'added' is a set containing new values that have been added.
+        And 'removed' is a set containing old values that have been removed.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        self = super().__new__(cls)
+        self.item_validator = accept_anything
+        self.notifiers = []
+        return self
+
+    def __init__(self, value=(), *, item_validator=None, notifiers=None):
+        if item_validator is not None:
+            self.item_validator = item_validator
+        super().__init__(self.item_validator(item) for item in value)
+        if notifiers is not None:
+            self.notifiers = notifiers
+
+    def notify(self, removed, added):
+        """ Call all notifiers.
+
+        This simply calls all notifiers provided by the class, if any.
+        The notifiers are expected to have the signature::
+
+            notifier(removed, added)
+
+        Any return values are ignored. Any exceptions raised are not
+        handled. Notifiers are therefore expected not to raise any
+        exceptions under normal use.
+
+        Parameters
+        ----------
+        removed : set
+            The items that have been removed.
+        added : set
+            The new items that have been added to the set.
+        """
+        for notifier in self.notifiers:
+            notifier(removed, added)
+
+    # -- set interface -------------------------------------------------------
+
+    def __iand__(self, value):
+        """  Return self &= value.
+
+        Parameters
+        ----------
+        value : set or frozenset
+            A value.
+
+        Returns
+        -------
+        self : TraitSet
+            The updated set.
+        """
+
+        old_set = self.copy()
+        retval = super().__iand__(value)
+        removed = old_set.difference(self)
+
+        if len(removed) > 0:
+            self.notify(removed, set())
+
+        return retval
+
+    def __ior__(self, value):
+        """ Return self |= value.
+
+        Parameters
+        ----------
+        value : set or frozenset
+            A value.
+
+        Returns
+        -------
+        self : TraitSet
+            The updated set.
+        """
+        old_set = self.copy()
+
+        if isinstance(value, (set, frozenset)):
```

Comment:
I suggest adding a comment immediately above this indicating why the check is necessary. Even knowing the context, my first thought was that we could just remove this line. It's not good that we're baking in assumptions about what Python currently does here (and those assumptions are fine right now, but may not be in the future), but I'm not seeing an alternative.

Reply:
Added the comment for now.